allennlp/training/trainer.py

```diff
@@ -360,7 +360,27 @@ def _train_epoch(self, epoch: int) -> Dict[str, float]:
         logger.info("Training")
 
         cumulative_batch_group_size = 0
+        stopped_early = False
         for batch_group in batch_group_generator_tqdm:
+            if self._distributed:
+                # Check whether the other workers have stopped already (due to differing amounts of
+                # data in each). If so, we can't proceed because we would hang when we hit the
+                # barrier implicit in Model.forward. We use a IntTensor instead a BoolTensor
+                # here because NCCL process groups apparently don't support BoolTensor.
+                done = torch.tensor(0, device=self.cuda_device)
+                torch.distributed.all_reduce(done, torch.distributed.ReduceOp.SUM)
+                if done.item() > 0:
```

Comment:
Why do they all have to stop at the same time? Is there no way of getting each worker to finish their share of the data, and then be done?

Reply:
Not that I know of, easily... I could imagine finishing early, grabbing the weights from the DDP model, creating a new model restricted to the workers that still had data and restarting, but that's going to be pretty gnarly. Definitely out of scope for this PR.

Reply:
The worker without data could just call `barrier()` a bunch until all workers signal that they're done. But let's fix that next time.

Reply:
> The worker without data could just call `barrier()` a bunch until all workers signal that they're done. But let's fix that next time.

I suggest, at least in the `_validation_loss` function of the trainer, we can use the `barrier()` logic instead of this `done_early` logic? Since if we drop some samples during the validation phase, the val result will be incorrect.